examples/platforms/efr32mg12/radio.c

```diff
@@ -0,0 +1,882 @@
+
+    RAIL_Idle(gRailHandle, RAIL_IDLE_ABORT, true); // abort packages under reception
+    sState = OT_RADIO_STATE_SLEEP;
+
+exit:
+    return error;
+}
+
+otError otPlatRadioReceive(otInstance *aInstance, uint8_t aChannel)
+{
+    otError          error = OT_ERROR_NONE;
+    RAIL_Status_t    status;
+    efr32BandConfig *config;
+
+    OT_UNUSED_VARIABLE(aInstance);
+    otEXPECT_ACTION(sState != OT_RADIO_STATE_DISABLED, error = OT_ERROR_INVALID_STATE);
+
+    config = efr32RadioGetBandConfig(aChannel);
+    otEXPECT_ACTION(config != NULL, error = OT_ERROR_INVALID_ARGS);
+
+    if (sCurrentBandConfig != config)
+    {
+        RAIL_Idle(gRailHandle, RAIL_IDLE_ABORT, true);
+        efr32RailConfigLoad(config);
+        sCurrentBandConfig = config;
+    }
+
+    status = RAIL_StartRx(gRailHandle, aChannel, NULL);
+    otEXPECT_ACTION(status == RAIL_STATUS_NO_ERROR, error = OT_ERROR_FAILED);
+
+    otLogInfoPlat("State=OT_RADIO_STATE_RECEIVE", NULL);
+    sState                 = OT_RADIO_STATE_RECEIVE;
+    sReceiveFrame.mChannel = aChannel;
+
+exit:
+    return error;
+}
+
+otError otPlatRadioTransmit(otInstance *aInstance, otRadioFrame *aFrame)
+{
+    otError           error      = OT_ERROR_NONE;
+    RAIL_CsmaConfig_t csmaConfig = RAIL_CSMA_CONFIG_802_15_4_2003_2p4_GHz_OQPSK_CSMA;
+    RAIL_TxOptions_t  txOptions  = RAIL_TX_OPTIONS_DEFAULT;
+    efr32BandConfig * config;
+    RAIL_Status_t     status;
+
+    if (sTransmitBusy)
+    {
+        // If the radio is already transmitting something, dont attempt to send a new packet.
+        // Let the upper layer know about the failure immediately.
+        otPlatRadioTxDone(aInstance, aFrame, NULL, OT_ERROR_BUSY);
+    }
+    else
+    {
+        otEXPECT_ACTION((sState != OT_RADIO_STATE_DISABLED) && (sState != OT_RADIO_STATE_TRANSMIT),
+                        error = OT_ERROR_INVALID_STATE);
+
+        config = efr32RadioGetBandConfig(aFrame->mChannel);
+        otEXPECT_ACTION(config != NULL, error = OT_ERROR_INVALID_ARGS);
+
+        sState         = OT_RADIO_STATE_TRANSMIT;
+        sTransmitError = OT_ERROR_NONE;
+        sTransmitBusy  = true;
+
+        if (sCurrentBandConfig != config)
+        {
+            RAIL_Idle(gRailHandle, RAIL_IDLE_ABORT, true);
+            efr32RailConfigLoad(config);
+            sCurrentBandConfig = config;
+        }
+
+        RAIL_WriteTxFifo(gRailHandle, &aFrame->mLength, sizeof(aFrame->mLength), true);
+        RAIL_WriteTxFifo(gRailHandle, aFrame->mPsdu, aFrame->mLength - 2, false);
+
+        if (aFrame->mPsdu[0] & IEEE802154_ACK_REQUEST)
+        {
+            txOptions |= RAIL_TX_OPTION_WAIT_FOR_ACK;
+        }
+
+        if (aFrame->mInfo.mTxInfo.mCsmaCaEnabled)
+        {
+            status = RAIL_StartCcaCsmaTx(gRailHandle, aFrame->mChannel, txOptions, &csmaConfig, NULL);
+        }
+        else
+        {
+            status = RAIL_StartTx(gRailHandle, aFrame->mChannel, txOptions, NULL);
```

Comment:
Is it possible for `RAIL_StartTx()` to return something other than `RAIL_STATUS_NO_ERROR` if the radio is busy receiving a packet or transmitting an acknowledgment?

Reply:
Yes it will send RADIO_STATUS_INVALID_STATE I think. We are adding some code to handle this..